lib/vkloader.c

```diff
@@ -146,38 +169,178 @@ ktxVulkanDeviceInfo_Construct(ktxVulkanDeviceInfo* This,
                               VkQueue queue, VkCommandPool cmdPool,
                               const VkAllocationCallbacks* pAllocator)
 {
-    VkCommandBufferAllocateInfo cmdBufInfo = {
-        .sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO
-    };
-    VkResult result;
+   return ktxVulkanDeviceInfo_ConstructEx(This, VK_NULL_HANDLE, physicalDevice, device, queue,
+                                          cmdPool, pAllocator, NULL);
+}
 
-#if defined(KTX_USE_FUNCPTRS_FOR_VULKAN)
-    // Delay loading not supported so must do it ourselves.
-    if (!ktxVulkanModuleHandle) {
-        ktx_error_code_e kresult = ktxLoadVulkanLibrary();
-        if (kresult != KTX_SUCCESS)
-            return kresult;
-    }
-#endif
+/**
+ * @memberof ktxVulkanDeviceInfo
+ * @~English
+ * @brief Construct a ktxVulkanDeviceInfo object.
+ *
+ * Records the device information, allocates a command buffer that will be
+ * used to transfer image data to the Vulkan device and retrieves the physical
+ * device memory properties for ease of use when allocating device memory for
+ * the images.
+ *
+ * Pass a valid ktxVulkanDeviceInfo* to any Vulkan KTX image loading
+ * function to provide it with the information.
+ *
+ * @param  This            pointer to the ktxVulkanDeviceInfo object to
+ *                        initialize.
+ * @param  physicalDevice handle of the Vulkan physical device.
+ * @param  device         handle of the Vulkan logical device.
+ * @param  queue          handle of the Vulkan queue.
+ * @param  cmdPool        handle of the Vulkan command pool.
+ * @param  pAllocator     pointer to the allocator to use for the image
+ *                        memory. If NULL, the default allocator will be used.
+ * @param  pFunctions     pointer to the struct of functions to use for vulkan
+ *                        operations.
+ *
+ * @returns KTX_SUCCESS on success, KTX_OUT_OF_MEMORY if a command buffer could
+ *          not be allocated.
+ *
+ * @sa ktxVulkanDeviceInfo_destruct()
+ */
+KTX_error_code
+ktxVulkanDeviceInfo_ConstructEx(ktxVulkanDeviceInfo* This,
+                              VkInstance instance,
+                              VkPhysicalDevice physicalDevice, VkDevice device,
+                              VkQueue queue, VkCommandPool cmdPool,
+                              const VkAllocationCallbacks* pAllocator,
+                              const ktxVulkanFunctions* pFunctions)
+{
 
+#define LOAD_EXT_FUNC(member, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)ktxLoadVulkanFunction(#fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
```

Comment:
Why do you have the `do { } while (0)` around this and the 2 macros below?

Reply:
This is just a general precausion: https://gcc.gnu.org/onlinedocs/cpp/Swallowing-the-Semicolon.html

Reply:
Seems to me it's six of one and half-a-dozen of the other. (That means it is the same either way. Awareness is needed.) With this you always have to put a semicolon after the macro call.

But okay, I usually write semicolons after macro calls so I'll go with this.